apps/keyboard/test/marionette/auto_correction_test.js

```diff
@@ -0,0 +1,210 @@
+
+      // Toggle again
+      keyboard.clickAutocorrectOption();
+      assert.equal(keyboard.autocorrect, !!currentSwitch);
+    });
+
+    test('re-open built-in settings', function() {
+      var currentSwitch = keyboard.autocorrect;
+
+      keyboard.clickAutocorrectOption();
+      keyboard.goBackToSettingsApp();
+      settings.close();
+
+      assert.equal(keyboard.autocorrect, !currentSwitch);
+
+      keyboard.launch();
+      keyboard.switchToBuiltInSettings();
+      keyboard.clickAutocorrectOption();
+
+      assert.equal(keyboard.autocorrect, !!currentSwitch);
+    });
+  });
+
+  suite('with autocorrect on', function() {
+    setup(function () {
+      if (!keyboard.autocorrect) {
+        keyboard.launch();
+        keyboard.switchToBuiltInSettings();
+        keyboard.clickAutocorrectOption();
+        keyboard.goBackToSettingsApp();
+        settings.close();
+      }
+
+      keyboardTestApp.launch();
+      keyboardTestApp.switchTo();
+      keyboardTestApp.textInput.tap();
+
+      keyboard.switchTo();
+    });
+
+    test('predictive tests', function() {
+      // Test 1 - type first 7 letters of the expected word
+      var expectedWord = 'keyboard';
+      keyboard.type(expectedWord.slice(0, 7));
+      keyboard.tapFirstPredictiveWord();
+      keyboardTestApp.switchTo();
+
+      assert.equal(
+        keyboardTestApp.textInput.getAttribute('value'), expectedWord);
+
+      // Test 2 - tap second suggestion, then press space
+      keyboard.switchTo();
+      keyboard.type(' ');
+      keyboard.type('Tes');
+      keyboardTestApp.switchTo();
+
+      assert.equal(
+        keyboardTestApp.textInput.getAttribute('value'), 'keyboard Tes');
+
+      keyboard.switchTo();
+      keyboard.tapSuggestionKey('Tea');
+      keyboard.type(' ');
+      keyboardTestApp.switchTo();
+
+      assert.equal(
+        keyboardTestApp.textInput.getAttribute('value'), 'keyboard Tea ');
+
+      // Test 3 - type something with autocorrect and press space
+      keyboard.switchTo();
+      keyboard.type('ye ');
+      keyboardTestApp.switchTo();
+
+      assert.equal(
+        keyboardTestApp.textInput.getAttribute('value'), 'keyboard Tea he ');
+
+      // Test 4 - autocorrect, dot and backspace
+      keyboard.switchTo();
+      keyboard.type('wot.');
+      keyboardTestApp.switchTo();
+
+      assert.equal(
+        keyboardTestApp.textInput.getAttribute('value'),
+        'keyboard Tea he wit.');
+
+      keyboard.switchTo();
+      keyboard.tapBackspaceKey();
+      keyboardTestApp.switchTo();
+
+      assert.equal(
+        keyboardTestApp.textInput.getAttribute('value'), 'keyboard Tea he wot');
+    });
+
+    test('bug 1073870', function() {
+      keyboard.type('Hello worl');
+      keyboardTestApp.switchTo();
+
+      client.executeScript(
+        'var el = document.activeElement;' +
+        'el.selectionStart = el.selectionEnd = 5;');
+
+      keyboard.switchTo();
+      keyboard.type(' ');
```

Comment:
You might need to wait for suggestion to disappear before typing .. this is racey.